### DISCUSSION TITLE
code editor

### DISCUSSION THREAD

**Discussion by nrtech1** [pybricks/support#1610] [General]
Hi sorry I am new to pybricks,
is there a code editor or is it just graphical programming?

**Comment by laurensvalk** [comment]
You can choose between block-based coding or Python coding. You can also mix the two, if you like.

When coding with blocks, you can also see the Python code being generated live as you move blocks around.

![image](https://github.com/pybricks/support/assets/12326241/b8e39880-37e4-4a74-8aad-c50132547654)

**Reply by nrtech1** [reply]
Lauren much appreciated, but when I try to edit code I get the message that I am in a read-only mode.
Thank you

**Reply by laurensvalk** [reply]
To work in Python, open a new Python project.

The generated code next to the blocks is read-only because otherwise you would lose all your hard work if you moved a blocks around and the code is regenerated right away :)

**Comment by nrtech1** [comment]
Lauren my apologies for the delay, thank you for your help.
regards

**Reply by laurensvalk** [reply]
No worries! You can pick either way, or even just use the blocks to do the boilerplate stuff for bigger Python programs :)

**Comment by nrtech1** [comment]
Lauren Hi
quick question for you. When I import a library in PyBricks and run it once, does it become available to another program?

**Reply by laurensvalk** [reply]
Imported libraries are re-downloaded along with your main program every time you run your main program.

So if you run another program, and you import a library there too, it is again included.

**Comment by nrtech1** [comment]
Lauren HiThank you for your reply. My apologies but I am still at the starting point of all of this. What I am trying to use is this PUPRemote.py
Its a library to communicate as a Powered UP (PUP) Device with LEGO smart hubs. It has a basic PUPDeviceEmulator and a higher-level PUPRemote library that acts more like RPC (Remote Procedure Calling).

So if I import this library and as long as its stays in the list of imported files, it will always be available to my main program?


Thank you for everythingNick

    On Wednesday, May 8, 2024 at 02:20:09 a.m. EDT, laurensvalk ***@***.***> wrote:  
 
 


Imported libraries are re-downloaded along with your main program every time you run your main program.

So if you run another program, and you import a library there too, it is again included.
